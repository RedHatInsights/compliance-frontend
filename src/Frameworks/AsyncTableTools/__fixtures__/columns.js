import { severityLevels } from './items';

export default [
  {
    title: 'ID',
    sortByProperty: 'id',
    exportKey: 'id',
    key: 'id',
    manageable: false,
  },
  {
    title: 'Name',
    key: 'id',
    sortByProperty: 'name',
    exportKey: 'name',
    renderFunc: (_a, _b, item) => item.name,
  },
  {
    title: 'Another Name column',
    key: 'another_name_column',
    sortByProperty: 'name',
    exportKey: 'name',
    renderExport: (name) => `${name} via export render`,
  },
  {
    title: 'Description',
    key: 'desc',
    sortByProperty: 'description',
    renderExport: (item) => `${item.name} description rendered for export`,
  },
  {
    title: 'Severity sorted by function',
    sortByFunction: (item) => item.name,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Severity by Array',
    sortByProperty: 'severity',
    sortByArray: severityLevels,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Download button',
    key: 'download',
    manageable: false,
    export: false,
    renderFunc: () => 'download link',
  },
];
