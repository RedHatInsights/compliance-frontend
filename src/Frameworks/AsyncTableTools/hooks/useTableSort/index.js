export { default } from './useTableSort';
export * from './useTableSort';
