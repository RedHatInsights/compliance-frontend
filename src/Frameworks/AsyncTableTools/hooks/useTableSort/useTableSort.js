import { useCallback, useMemo } from 'react';
import { addSortableTransform, columnOffset } from './helpers';
import useTableState from '../useTableState';
import { TABLE_STATE_NAMESPACE } from './constants';

/**
 *  @typedef {object} useTableSortReturn
 *
 *  @property {object}   [tableProps]         Props for a Patternfly table
 *  @property {Function} [tableProps.onSort]  Callback function for column headers in a Patternfly table
 *  @property {Array}    [tableProps.cells]   Array containing columns for a Patternfly table with the sortable transform applied
 *  @property {object}   [tableProps.sortBy ] Object containing the current sortBy state
 *
 */

/**
 * Provides columns with the `sortable` transform mixed in for a Patternfly table.
 *
 *  @param   {Array}              columns                    Columns for a table, with a "sortable" prop
 *  @param   {object}             [options]                  AsyncTableTools options
 *  @param   {object}             [options.sortBy]           An initial sortBy state like `{index: 1, direction: 'desc'}`
 *  @param   {object}             [options.onSort]           A function to call after setting a new sort state.
 *  @param   {object}             [options.serialisers.sort] A function to provide a serialiser for the table state
 *
 *  @returns {useTableSortReturn}                            Props for a Patternfly table to integrate sorting
 *
 *  @example
 *
 * const columns = [{ title: 'Name', sortable: true }]
 * const tableSort = useTableSort(columns)
 *
 *  @category AsyncTableTools
 *  @subcategory Hooks
 *
 */
const useTableSort = (columns, options = {}) => {
  const {
    sortBy: initialSortBy,
    serialisers,
    onSelect,
    detailsComponent,
    onSort: onSortOption,
  } = options;
  const defaultSortBy = useMemo(
    () =>
      initialSortBy || {
        index: columnOffset({
          detailsComponent,
          onSelect,
        }),
        direction: 'asc',
      },
    [detailsComponent, onSelect, initialSortBy]
  );

  const serialiser = useCallback(
    (state) => options.serialisers.sort(state, columns),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [JSON.stringify(columns), JSON.stringify(options.serialisers)]
  );
  const stateOptions = useMemo(
    () => ({
      ...(serialisers?.sort
        ? {
            serialiser,
          }
        : {}),
    }),
    [serialisers, serialiser]
  );
  const [sortBy, setSortBy] = useTableState(
    TABLE_STATE_NAMESPACE,
    defaultSortBy,
    stateOptions
  );

  const onSort = useCallback(
    (_, index, direction) => {
      setSortBy({
        index,
        direction,
      });
      onSortOption?.(index, direction);
    },
    [onSortOption, setSortBy]
  );

  return {
    tableProps: {
      onSort,
      sortBy: sortBy || defaultSortBy,
      cells: addSortableTransform(columns),
    },
  };
};

export default useTableSort;
