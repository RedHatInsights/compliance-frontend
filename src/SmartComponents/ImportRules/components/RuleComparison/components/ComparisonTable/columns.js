import React from 'react';
import { nowrap } from '@patternfly/react-table';

import RuleCell from './components/RuleCell';
import ProfileVersionCell from './components/ProfileVersionCell';

const rule = {
  title: <strong>Rule</strong>,
  Component: RuleCell,
};

export const profileVersion = (version, targetVersion, additionalProps) => ({
  dataLabel: `RHEL ${version.os_major_version}.${version.os_minor_version} -  SSG version ${version.ssg_version}`,
  title: (
    <>
      <strong>
        RHEL {version.os_major_version}.{version.os_minor_version}
      </strong>
      <br />
      SSG version {version.ssg_version}
    </>
  ),
  Component: (props) => (
    <ProfileVersionCell
      {...{
        version,
        targetVersion,
        ...props,
        ...additionalProps,
      }}
    />
  ),
  transforms: [nowrap],
});

export default [rule];
