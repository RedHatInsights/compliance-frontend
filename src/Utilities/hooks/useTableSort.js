import { useState } from 'react';
import { orderArrayByProp } from 'Utilities/helpers';

const useTableSort = (array, columns) => {
    const [sortBy, setSortBy] = useState({
        index: 0,
        direction: 'desc'
    });
    const column = columns[sortBy.index];
    const onSort = (_, index, direction) => (
        setSortBy({
            index,
            direction
        })
    );

    return {
        tableSort: {
            onSort,
            sortBy
        },
        sorted: orderArrayByProp(
            (column?.sortByProperty || column?.sortByFunction), array, sortBy.direction
        )
    };
};

export default useTableSort;
