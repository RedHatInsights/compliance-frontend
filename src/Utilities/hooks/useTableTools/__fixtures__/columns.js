import { severityLevels } from './items';

export default [
    {
        title: 'Name',
        sortByProperty: 'name',
        renderFunc: (_a, _b, item) => (
            item.name
        )
    }, {
        title: 'Description',
        sortByProperty: 'description'
    }, {
        title: 'Severity sorted by function',
        sortByFunction: (item) => (item.name),
        renderFunc: (_a, _b, item) => (
            item.severity
        )
    }, {
        title: 'Severity by Array',
        sortByProperty: 'severity',
        sortByArray: severityLevels,
        renderFunc: (_a, _b, item) => (
            item.severity
        )
    }
];
