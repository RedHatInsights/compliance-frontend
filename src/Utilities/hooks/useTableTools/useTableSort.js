import { useState, useCallback } from 'react';
import { sortable } from '@patternfly/react-table';
import { orderArrayByProp, orderByArray, uniq } from 'Utilities/helpers';

const isSortable = (column) => column.sortByProp || column.sortByFunction;

const addSortableTransform = (columns) =>
  columns.map((column) => ({
    ...column,
    ...(isSortable(column)
      ? {
          transforms: uniq([...(column?.transforms || []), sortable]),
        }
      : {}),
  }));

const columnOffset = (options = {}) =>
  (typeof options.onSelect === 'function' || options.hasRadioSelect ? 1 : 0) +
  (typeof options.detailsComponent !== 'undefined');

const sortByFromOptions = (options = {}) => ({
  index:
    options.tableType === 'tree'
      ? options.sortBy?.index - 1
      : options.sortBy?.index || 1,
  direction: options.sortBy?.direction || 'asc',
});

const useTableSort = (columns, options = {}) => {
  const [sortBy, setSortBy] = useState(sortByFromOptions(options));
  const onSort = (_, index, direction) => {
    setSortBy({
      index,
      direction,
    });
  };

  const sorter = useCallback(
    (items) => {
      const currentSortableColumn =
        columns[
          sortBy.index -
            (options.tableType === 'tree' ? 0 : columnOffset(options)) -
            1
        ];
      return currentSortableColumn?.sortByArray
        ? orderByArray(
            items,
            currentSortableColumn?.sortByProp,
            currentSortableColumn?.sortByArray,
            sortBy.direction
          )
        : orderArrayByProp(
            currentSortableColumn?.sortByProp ||
              currentSortableColumn?.sortByFunction,
            items,
            sortBy.direction
          );
    },
    [sortBy, columns, options.tableType]
  );

  return {
    sorter,
    tableProps: {
      onSort,
      sortBy,
      cells: addSortableTransform(columns),
    },
  };
};

export const useTableSortWithItems = (items, columns, options) => {
  const { tableProps, sorter } = useTableSort(columns, options);

  return {
    tableProps: {
      ...tableProps,
      sortBy: items.length > 0 ? tableProps.sortBy : undefined,
    },
    sorter,
  };
};

export default useTableSort;
