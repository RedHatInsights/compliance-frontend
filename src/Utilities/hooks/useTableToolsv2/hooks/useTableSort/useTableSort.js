import { addSortableTransform, columnOffset } from './helpers';
import useTableState from '../useTableState';

// TODO implement onSetSortState to transform state put into the tablestate to be API consumable
/**
 * Provides columns with the `sortable` transform mixed in for a Patternfly table.
 *
 * @param {Array} columns Columns for a table
 * @param {Object} [options]
 */
const useTableSort = (columns, options = {}) => {
  const [sortBy, setSortBy] = useTableState(
    'sort',
    options.sortBy || {
      index: columnOffset(options),
      direction: 'asc',
    }
  );
  const onSort = (_, index, direction) => {
    setSortBy({
      index,
      direction,
    });
    options.onSort?.(index, direction);
  };

  return {
    sortBy,
    tableProps: {
      onSort,
      sortBy,
      cells: addSortableTransform(columns),
    },
  };
};

export default useTableSort;
