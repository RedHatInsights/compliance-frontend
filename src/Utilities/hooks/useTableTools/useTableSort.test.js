import { renderHook } from '@testing-library/react';
import useTableSort, { useTableSortWithItems } from './useTableSort';
import columns from './__fixtures__/columns';

describe('useTableSort', () => {
  it('returns a table sort configuration', () => {
    const { result } = renderHook(() => useTableSort(columns));
    expect(result).toMatchSnapshot();
  });

  it('returns a table sort configuration with an inital state', () => {
    const sortBy = {
      index: 3,
      direction: 'asc',
    };
    const { result } = renderHook(() =>
      useTableSort(columns, {
        sortBy,
      })
    );
    expect(result.current.tableProps.sortBy).toEqual(sortBy);
  });
});

describe('useTableSortWithItems', () => {
  it('returns no sortBy when there are no items', () => {
    const sortBy = {
      index: 3,
      direction: 'asc',
    };
    const { result } = renderHook(() =>
      useTableSortWithItems([], columns, {
        sortBy,
      })
    );
    expect(result.current.tableProps.sortBy).toEqual(undefined);
  });
});
