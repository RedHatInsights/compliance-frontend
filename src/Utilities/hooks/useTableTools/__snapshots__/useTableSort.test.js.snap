// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTableSort returns a table sort configuration 1`] = `
{
  "current": {
    "sorter": [Function],
    "tableProps": {
      "cells": [
        {
          "exportKey": "id",
          "key": "id",
          "managable": false,
          "sortByProperty": "id",
          "title": "ID",
        },
        {
          "exportKey": "name",
          "key": "id",
          "renderFunc": [Function],
          "sortByProperty": "name",
          "title": "Name",
        },
        {
          "exportKey": "name",
          "key": "another_name_column",
          "renderExport": [Function],
          "sortByProperty": "name",
          "title": "Another Name column",
        },
        {
          "key": "desc",
          "renderExport": [Function],
          "sortByProperty": "description",
          "title": "Description",
        },
        {
          "exportKey": "severity",
          "renderFunc": [Function],
          "sortByFunction": [Function],
          "title": "Severity sorted by function",
          "transforms": [
            [Function],
          ],
        },
        {
          "exportKey": "severity",
          "renderFunc": [Function],
          "sortByArray": [
            "high",
            "medium",
            "low",
            "unknown",
          ],
          "sortByProperty": "severity",
          "title": "Severity by Array",
        },
      ],
      "onSort": [Function],
      "sortBy": {
        "direction": "asc",
        "index": 1,
      },
    },
  },
}
`;
